compiler/grainformat/reformat.re

```diff
@@ -1045,78 +1055,87 @@ and print_application =
             ~endChar=None,
             ~original_source,
             ~parent_loc,
+            ~level=0,
             first,
           ),
           Doc.rparen,
         ])
-      | PExpApp(fn, _) =>
-        let leftfn = get_function_name(fn);
-        if (infixop(leftfn)) {
-          Doc.concat([
-            Doc.lparen,
-            print_expression(
-              ~parentIsArrow=false,
-              ~endChar=None,
-              ~original_source,
-              ~parent_loc,
-              first,
-            ),
-            Doc.rparen,
-          ]);
-        } else {
-          Doc.concat([
-            Doc.lparen,
-            print_expression(
-              ~parentIsArrow=false,
-              ~endChar=None,
-              ~original_source,
-              ~parent_loc,
-              first,
-            ),
-            Doc.rparen,
-          ]);
-        };
+      | PExpApp(_) =>
+        print_expression(
+          ~parentIsArrow=false,
+          ~endChar=None,
+          ~original_source,
+          ~parent_loc,
+          ~level=1,
+          first,
+        )
+
       | _ =>
         print_expression(
           ~parentIsArrow=false,
           ~endChar=None,
           ~original_source,
           ~parent_loc,
+          ~level=0,
           first,
         )
       };
 
     let second_brackets =
       switch (second.pexp_desc) {
-      | PExpIf(_)
-      | PExpApp(_) =>
+      | PExpIf(_) =>
         Doc.concat([
           Doc.lparen,
           print_expression(
             ~parentIsArrow=false,
             ~endChar=None,
             ~original_source,
             ~parent_loc,
+            ~level=0,
             second,
           ),
           Doc.rparen,
         ])
+      | PExpApp(_) =>
+        print_expression(
+          ~parentIsArrow=false,
+          ~endChar=None,
+          ~original_source,
+          ~parent_loc,
+          ~level=1,
+          second,
+        )
       | _ =>
         print_expression(
           ~parentIsArrow=false,
           ~endChar=None,
           ~original_source,
           ~parent_loc,
+          ~level=0,
           second,
         )
       };
-    Doc.concat([
-      first_brackets,
-      Doc.space,
-      Doc.text(function_name),
-      Doc.space,
-      second_brackets,
-    ]);
+
+    if (level > 0 && !has_standard_precedence(function_name)) {
+      // only wrap maths functions (where there's a precedence)
```

Comment:
This logic isn't _quite_ right, since you're welcome to shadow any of the operators but they keep their precedence (`++` has the same precedence as `+`):
```
let (++) = (a, b) => a + a + b + b
```

Reply:
Ah, good point.

Reply:
I was going to do proper precedence comparison anyway to determine if the brackets were needed.   Any  thoughts on keeping track of that for shadowed operators?

Reply:
No need to keep track! The precedence remains the same regardless of whether you shadow or not.

Reply:
Converting this to Draft whilst I implement this fix

Reply:
I added a naieve implementation if you could take a look?